File: Wire-iOS/Sources/Managers/ProximityMonitorManager.m
// 
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 


#import "ProximityMonitorManager.h"
#import "WireSyncEngine+iOS.h"
#import "Constants.h"
#import "ZClientViewController.h"
#import "avs+iOS.h"
#import "Settings.h"
#import "Wire-Swift.h"


@interface ProximityMonitorManager() <VoiceChannelStateObserver, AVSMediaManagerClientObserver>

@property (nonatomic) id voiceChannelStateObserverToken;
@property (nonatomic) VoiceChannelV2State lastVoiceChannelState;


@end



@implementation ProximityMonitorManager

- (id)init
{
    self = [super init];
    
    if (self) {
        self.voiceChannelStateObserverToken = [VoiceChannelV3 addStateObserver:self userSession:[ZMUserSession sharedSession]];


        // Wait and then update initial state so that getting the voice channel doesn't affect startup time
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(1.0 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
            ZMConversation *activeCallConversation = [WireCallCenter activeCallConversationsInUserSession:[ZMUserSession sharedSession]].firstObject;
            self.lastVoiceChannelState = activeCallConversation.voiceChannel.state;
            [self updateProximityMonitorState];
        });
        
        if (![[Settings sharedSettings] disableAVS]) {
            [AVSMediaManagerClientChangeNotification addObserver:self];
        }
    }
    return self;
}

- (void)dealloc
{
    if (![[Settings sharedSettings] disableAVS]) {
        [AVSMediaManagerClientChangeNotification removeObserver:self];
    }
}

- (void)updateProximityMonitorState
{
    // Don't use proximity monitoring on the ipad
    if (IS_IPAD) {
        return;
    }
    
    if (self.lastVoiceChannelState == VoiceChannelV2StateOutgoingCall) {
        [UIDevice currentDevice].proximityMonitoringEnabled = YES;
    }
    else {
        BOOL isConnectingOrActive = self.lastVoiceChannelState == VoiceChannelV2StateSelfConnectedToActiveChannel ||
                                    self.lastVoiceChannelState == VoiceChannelV2StateSelfIsJoiningActiveChannel ||
                                    self.lastVoiceChannelState == VoiceChannelV2StateOutgoingCallInactive;
        
        if (isConnectingOrActive && [[AVSProvider shared] mediaManager].speakerEnabled == NO) {
            [UIDevice currentDevice].proximityMonitoringEnabled = YES;
        }
        else {
            [UIDevice currentDevice].proximityMonitoringEnabled = NO;
        }
    }
}

#pragma mark - VoiceChannelStateObserver

- (void)callCenterDidChangeVoiceChannelState:(VoiceChannelV2State)voiceChannelState conversation:(ZMConversation *)conversation callingProtocol:(enum CallingProtocol)callingProtocol
{
    self.lastVoiceChannelState = voiceChannelState;
    [self updateProximityMonitorState];
}

- (void)callCenterDidFailToJoinVoiceChannelWithError:(NSError *)error conversation:(ZMConversation *)conversation
{
    
}

- (void)callCenterDidEndCallWithReason:(VoiceChannelV2CallEndReason)reason conversation:(ZMConversation *)conversation callingProtocol:(enum CallingProtocol)callingProtocol
{
    
}

#pragma mark - AVSMediaManagerClientObserver

- (void)mediaManagerDidChange:(AVSMediaManagerClientChangeNotification *)notification
{
    if (notification.speakerEnableChanged) {
        [self updateProximityMonitorState];
    }
}

@end

File: wire-ios/Wire-iOS/Sources/Managers/ProximityMonitorManager.swift
//
// Wire
// Copyright (C) 2024 Wire Swiss GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
//

import avs
import UIKit
import WireDataModel
import WireSyncEngine

private let zmLog = ZMSLog(tag: "calling")

final class ProximityMonitorManager: NSObject {
    typealias RaisedToEarHandler = (_ raisedToEar: Bool) -> Void

    var callStateObserverToken: Any?

    fileprivate(set) var raisedToEar = false {
        didSet {
            if oldValue != raisedToEar {
                stateChanged?(raisedToEar)
            }
        }
    }

    var stateChanged: RaisedToEarHandler?
    var listening = false

    deinit {
        AVSMediaManagerClientChangeNotification.remove(self)
        self.stopListening()
        NotificationCenter.default.removeObserver(self)
    }

    override init() {
        super.init()

        guard let userSession = ZMUserSession.shared() else {
            zmLog.error("UserSession not available when initializing \(type(of: self))")
            return
        }

        self.callStateObserverToken = WireCallCenterV3.addCallStateObserver(observer: self, userSession: userSession)
        AVSMediaManagerClientChangeNotification.add(self)

        updateProximityMonitorState()
    }

    func updateProximityMonitorState() {
        // Only do proximity monitoring on phones
        guard UIDevice.current.userInterfaceIdiom == .phone, let callCenter = ZMUserSession.shared()?.callCenter,
              !listening else { return }

        let ongoingCalls = callCenter.nonIdleCalls.filter { (_, callState: CallState) -> Bool in
            switch callState {
            case .established, .establishedDataChannel, .answered(degraded: false), .outgoing(degraded: false):
                return true
            default:
                return false
            }
        }

        let hasOngoingCall = ongoingCalls.count > 0
        let speakerIsEnabled = AVSMediaManager.sharedInstance()?.isSpeakerEnabled ?? false

        UIDevice.current.isProximityMonitoringEnabled = !speakerIsEnabled && hasOngoingCall
    }

    // MARK: - listening mode switching (for AudioMessageView)

    func startListening() {
        guard !listening else {
            return
        }

        listening = true

        UIDevice.current.isProximityMonitoringEnabled = true
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(handleProximityChange),
            name: UIDevice.proximityStateDidChangeNotification,
            object: nil
        )
    }

    func stopListening() {
        guard listening else {
            return
        }
        listening = false

        UIDevice.current.isProximityMonitoringEnabled = false
    }

    @objc
    func handleProximityChange(_: Notification) {
        raisedToEar = UIDevice.current.proximityState
    }
}

extension ProximityMonitorManager: WireCallCenterCallStateObserver {
    func callCenterDidChange(
        callState: CallState,
        conversation: ZMConversation,
        caller: UserType,
        timestamp: Date?,
        previousCallState: CallState?
    ) {
        updateProximityMonitorState()
    }
}

extension ProximityMonitorManager: AVSMediaManagerClientObserver {
    func mediaManagerDidChange(_ notification: AVSMediaManagerClientChangeNotification!) {
        if notification.speakerEnableChanged {
            updateProximityMonitorState()
        }
    }
}
